src/iocp/mod.rs

```diff
@@ -0,0 +1,820 @@
+            }
+        };
+
+        // Indicate to the source that it is being deleted.
+        // This cancels any ongoing AFD_IOCTL_POLL operations.
+        source.begin_delete()
+    }
+
+    /// Wait for events.
+    pub(super) fn wait(&self, events: &mut Events, timeout: Option<Duration>) -> io::Result<()> {
+        log::trace!("wait: handle={:?}, timeout={:?}", self.port, timeout);
+
+        let deadline = timeout.and_then(|timeout| Instant::now().checked_add(timeout));
+        let mut packets = self.packets.lock().unwrap_or_else(|e| e.into_inner());
+        let mut notified = false;
+        events.packets.clear();
+
+        loop {
+            let mut new_events = 0;
+
+            // Indicate that we are now polling.
+            debug_assert!(!self.polling.swap(true, Ordering::SeqCst));
+
+            let guard = CallOnDrop(|| {
+                debug_assert!(self.polling.swap(false, Ordering::SeqCst));
+            });
+
+            // Process every entry in the queue before we start polling.
+            self.drain_update_queue(false)?;
+
+            // Get the time to wait for.
+            let timeout = deadline.map(|t| t.saturating_duration_since(Instant::now()));
+
+            // Wait for I/O events.
+            let len = self.port.wait(&mut packets, timeout)?;
+            log::trace!("new events: handle={:?}, len={}", self.port, len);
+
+            // We are no longer polling.
+            drop(guard);
+
+            // Process all of the events.
+            for entry in packets.drain(..) {
+                let packet = entry.into_packet();
+
+                // Feed the event into the packet.
+                match packet.feed_event(self)? {
+                    FeedEventResult::NoEvent => {}
+                    FeedEventResult::Event(event) => {
+                        events.packets.push(event);
+                        new_events += 1;
+                    }
+                    FeedEventResult::Notified => {
+                        notified = true;
+                    }
+                }
+            }
+
+            // Break if there was a notification or at least one event, or if deadline is reached.
+            let timeout_is_empty =
+                timeout.map_or(false, |t| t.as_secs() == 0 && t.subsec_nanos() == 0);
+            if notified || new_events > 0 || timeout_is_empty {
+                break;
+            }
+
+            log::trace!("wait: no events found, re-entering polling loop");
+        }
+
+        Ok(())
+    }
+
+    /// Notify this poller.
+    pub(super) fn notify(&self) -> io::Result<()> {
+        // Push the notify packet into the IOCP.
+        self.port.post(0, 0, self.notifier.clone())
+    }
+
+    /// Run an update on a packet.
+    fn update_packet(&self, mut packet: Packet) -> io::Result<()> {
+        loop {
+            // If we are currently polling, we need to update the packet immediately.
+            if self.polling.load(Ordering::Acquire) {
```

Comment:
so, will this then behave correctly on release builds, when `self.polling` never gets modified?